### PR TITLE
DAOS-15056 rebuild: add rpt to the rgt list properly

### PR DESCRIPTION
**Pull request by wangdi1**
Add rpt to the rgt list before yield, in case the rpt is incorrectly created with incoming scan RPC.

Features: rebuild
Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by liuxuezhao — DISMISSED** [review]

File: src/rebuild/srv.c [hunk 1]
```diff
@@ -2506,14 +2514,16 @@ rebuild_tgt_prepare(crt_rpc_t *rpc, struct rebuild_tgt_pool_tracker **p_rpt)
...
-	rpt_get(rpt);
-	d_list_add(&rpt->rt_list, &rebuild_gst.rg_tgt_tracker_list);
 	*p_rpt = rpt;
 out:
 	if (rc) {
-		if (rpt)
+		if (rpt) {
 			rpt_put(rpt);
-
+			if (!d_list_empty(&rpt->rt_list)) {
```
Comment (wangshilong) [inline src/rebuild/srv.c:2522]: This check better put before above rpt_put() to avoid accessing freed memory?

File: src/rebuild/srv.c [hunk 2]
```diff
@@ -2506,14 +2514,16 @@ rebuild_tgt_prepare(crt_rpc_t *rpc, struct rebuild_tgt_pool_tracker **p_rpt)
...
+				d_list_del_init(&rpt->rt_list);
+				rpt_put(rpt);
```
Comment (NiuYawei) [inline src/rebuild/srv.c:2524]: rpt_put() doesn't destroy 'rpt' when last reference dropped? I'm just wondering who is responsible for removing the 'rpt' from the global list and destroy it? It looks to me that there is no 'definite last reference holder' for 'rpt', right?

**Review by NiuYawei — CHANGES_REQUESTED** [review]
Please address the 'rpt' leak concern as well.

File: src/rebuild/srv.c (hunk 1 above)
Comment (wangdi1) [reply src/rebuild/srv.c:2527]: Hmm, actually if rpt is created successfully, then rpt must be in the list anyway. Though yes, it is better to put after rt_list indeed. thanks.

**Review by NiuYawei — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Pull request closed without merging**